controllers/watcherapi_controller.go

```diff
@@ -18,39 +18,237 @@ package controllers
-	// TODO(user): your logic here
+func (r *WatcherAPIReconciler) Reconcile(ctx context.Context, req ctrl.Request) (result ctrl.Result, _err error) {
+
+	Log := r.GetLogger(ctx)
+	instance := &watcherv1beta1.WatcherAPI{}
+	err := r.Client.Get(ctx, req.NamespacedName, instance)
+	if err != nil {
+		if k8s_errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers. Return and don't requeue.
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	Log.Info(fmt.Sprintf("Reconciling WatcherAPI instance '%s'", instance.Name))
+
+	helper, err := helper.NewHelper(
+		instance,
+		r.Client,
+		r.Kclient,
+		r.Scheme,
+		Log,
+	)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	isNewInstance := instance.Status.Conditions == nil
+	// Save a copy of the conditions so that we can restore the LastTransitionTime
+	// when a condition's state doesn't change.
+	savedConditions := instance.Status.Conditions.DeepCopy()
+
+	// Always patch the instance status when exiting this function so we can
+	// persist any changes.
+	defer func() {
+		condition.RestoreLastTransitionTimes(
+			&instance.Status.Conditions, savedConditions)
+		if instance.Status.Conditions.IsUnknown(condition.ReadyCondition) {
+			instance.Status.Conditions.Set(
+				instance.Status.Conditions.Mirror(condition.ReadyCondition))
+		}
+		err := helper.PatchInstance(ctx, instance)
+		if err != nil {
+			_err = err
+			return
+		}
+	}()
+
+	err = r.initStatus(instance)
+	if err != nil {
+		return ctrl.Result{}, nil
+	}
+
+	// If we're not deleting this and the service object doesn't have our finalizer, add it.
+	if instance.DeletionTimestamp.IsZero() && controllerutil.AddFinalizer(instance, helper.GetFinalizer()) || isNewInstance {
+		return ctrl.Result{}, nil
+	}
+
+	// Handle service delete
+	if !instance.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, instance, helper)
+	}
+
+	configVars := make(map[string]env.Setter)
+	// check for required OpenStack secret holding passwords for service/admin user and add hash to the vars map
+	Log.Info(fmt.Sprintf("[API] Get secret 1 '%s'", instance.Spec.Secret))
+	ctrlResult, err := r.getSecret(ctx, instance, helper, instance.Spec.Secret, []string{instance.Spec.PasswordSelectors.Service}, &configVars)
+	if err != nil {
+		return ctrlResult, err
+	}
+
+	err = r.generateServiceConfigs(ctx, instance, helper, &configVars)
+	if err != nil {
+		instance.Status.Conditions.Set(condition.FalseCondition(
+			condition.InputReadyCondition,
+			condition.ErrorReason,
+			condition.SeverityWarning,
+			condition.InputReadyErrorMessage,
+			err.Error()))
+		return ctrl.Result{}, err
+	}
+
+	// all our input checks out so report InputReady
+	instance.Status.Conditions.MarkTrue(condition.InputReadyCondition, condition.InputReadyMessage)
```

Comment:
I was thinking of leaving the ServiceConfigReadyCondition for later, because all the functionality is there, but I see how it can be confusing in the current state. I've added both conditions now